viewing_party/party.py

```diff
@@ -1,23 +1,121 @@
-# ------------- WAVE 4 --------------------
-# -----------------------------------------
+# ---------------------------------------------------- WAVE 2 ----------------------------------------------------
 
-# -----------------------------------------
-# ------------- WAVE 5 --------------------
-# -----------------------------------------
+def get_watched_avg_rating(user_data):
+    if not user_data["watched"]:
+        return 0
+    
+    total_movie_ratings = 0
 
+    for movie_rating in user_data["watched"]:
+        total_movie_ratings += movie_rating["rating"]
+
+    average_movie_rating = total_movie_ratings / len(user_data["watched"])
+
+    return average_movie_rating
+
+def get_most_watched_genre(user_data):
+    frequency_of_genre_index = {}
+    max_frequency = 0
+    most_watched_genre = None
+
+    for single_movie in user_data["watched"]:
+        genre_type = single_movie.get("genre")
+        if genre_type:
+            frequency_of_genre_index[genre_type] = frequency_of_genre_index.get(genre_type,0) + 1
+            if frequency_of_genre_index[genre_type] > max_frequency:
+              max_frequency = frequency_of_genre_index[genre_type]
+              most_watched_genre = genre_type
+
+    return most_watched_genre
+
+# ---------------------------------------------------- WAVE 3 ----------------------------------------------------
+
+def get_unique_watched(user_data):
+    if not user_data["watched"]:
+        return []
+  
+    my_friends_unique = [movie["title"] for my_friends in user_data["friends"] 
+                         for movie in my_friends["watched"]]
+    
+    my_unique_movies = [movie for movie in user_data["watched"] 
+                        if movie["title"] not in my_friends_unique]   
+
+    return my_unique_movies
+
+def get_friends_unique_watched(user_data):
+    friends_unique_movies = []
+    my_movie_titles = [] 
+    
+    my_movie_titles = [movie["title"] for movie in user_data["watched"]]
+
+    for my_friend in user_data["friends"]:           
+        for movie in my_friend["watched"]:
+            if movie["title"] not in my_movie_titles and movie not in friends_unique_movies:
+                friends_unique_movies.append(movie) 
+
+    return list(friends_unique_movies)
+
+# ---------------------------------------------------- WAVE 4 ----------------------------------------------------
+
+def get_available_recs(user_data):
+    recommended_movies = []
+    friends_unique_movies = get_friends_unique_watched(user_data)
+
+    user_watched_titles = set()
+    
+    user_watched_titles = [movie['title'] for movie in user_data.get('watched', [])]
+
+    user_subscriptions = set(user_data.get('subscriptions', []))
+
+    recommended_movies = [movie for movie in friends_unique_movies 
+                          if movie['host'] in user_subscriptions 
+                          and movie['title'] not in user_watched_titles 
+                          if movie not in recommended_movies]

```

Comment:
I think that this is a good example of maybe opting to not us list comprehension, you lose out on readability here. Even though not using comprehension takes us a little longer to type out, both ways function the same. We would re-gain the loss readability we lost here.